src/media.rs

```diff
@@ -32,78 +29,78 @@ pub fn generate_thumbnail_and_get_size(image: &[u8]) -> (Vec<u8>, (u32, u32)) {
-                let file_encrypted_hash = file_encrypted_hash.clone();
-                let file_encrypted = file_encrypted.clone();
-                let media_key = media_key.clone();
-                let callback = callback.clone();
-
-                thread::spawn(move || {
-                    let form = reqwest::multipart::Form::new()
-                        .text("hash", base64::encode(&file_encrypted_hash.to_vec()))
-                        .part("file", reqwest::multipart::Part::reader(Cursor::new(file_encrypted.to_vec()))
-                            .mime(reqwest::mime::APPLICATION_OCTET_STREAM));
-
-                    let file_info = reqwest::Client::new().post(url.as_str())
-                        .multipart(form)
-                        .send()
-                        .and_then(|mut response| response.text())
-                        .map_err(|e| Error::with_chain(e, "could not upload file"))
-                        .and_then(|response| json::parse(&response).map_err(|e| (Error::with_chain(e, "invalid response"))))
-                        .and_then(|json| json.get_str("url").map(|url| url.to_string()))
-                        .map(|url| FileInfo {
-                            mime: "image/jpeg".to_string(),
-                            sha256: file_hash.to_vec(),
-                            enc_sha256: file_encrypted_hash.to_vec(),
-                            key: media_key.to_vec(),
-                            url,
-                            size: file_len, //Or encrypted file size ??
-                        });
-                    callback(file_info);
-                });
-            }
-            Err(err) => callback(Err(err).map_err(|_| "could not request file upload"))
-        }
-    }))
+pub async fn upload_file(file: &[u8], mime: String, media_type: MediaType, auth: &str, host: &str) -> Result<FileInfo> {
+     let sha256 = crypto::sha256(file);
+
+    let (file_encrypted, key) = crypto::encrypt_media_message(media_type, file);
+    let enc_sha256 = crypto::sha256(&file_encrypted);
+    let token = base64::encode_config(&enc_sha256, base64::URL_SAFE_NO_PAD);
+    let size = file.len();
+    let mut url = reqwest::Url::parse(&format!("https://{}/", host)).unwrap();
+    url.path_segments_mut().unwrap()
+        .extend(&path_for(media_type))
+        .push(&token);
+
+    let client = reqwest::Client::builder()
+        .user_agent("curl/7.69.1")
+        .build()?;
+
+    eprintln!("{}", url);
+    let response = client
+        .post(&url.to_string())
+        .query(&[
+            ("auth", auth),
+            ("token", &token),
+        ])
+        .header("Origin", "https://web.whatsapp.com")
+        .header("Referer", "https://web.whatsapp.com/")
+        .body(file_encrypted)
+        .send().await?;
+
+
+    if response.status() != reqwest::StatusCode::from_u16(200).unwrap() {
+        error!("{:?}", response);
+        error!("{:?}", response.text().await?);
```

Comment:
(remove / rework these errors as well, as discussed above)

Reply:
The response data contains useful information what is not available on the returned Error object. Perhaps we should improve the Error enumeration or replace this with a debug! log.

Reply:
Either solution would be alright, though I'd prefer it if you improved the `Error` enum :)